lib/Parser/Parse.cpp

```diff
@@ -2447,6 +2523,146 @@ IdentPtr Parser::ParseImportOrExportFromClause(bool throwIfNotFound)
     return moduleSpecifier;
 }
 
+template<bool buildAST> 
+ParseNodePtr Parser::ParseDefaultExportClause()
+{
+    Assert(m_token.tk == tkDEFAULT);
+
+    m_pscan->Scan();
+    ParseNodePtr pnode = nullptr;
+    ushort flags = fFncNoFlgs;
+    
+    switch (m_token.tk)
+    {
+    case tkCLASS:
+        {
+            if (!m_scriptContext->GetConfig()->IsES6ClassAndExtendsEnabled())
+            {
+                goto LDefault;
+            }
+
+            // Before we parse the class itself we need to know if the class has an identifier name.
+            // If it does, we'll treat this class as an ordinary class declaration which will bind 
+            // it to that name. Otherwise the class should parse as a nameless class expression and 
+            // bind only to the export binding.
+            BOOL classHasName = false;
+            RestorePoint parsedClass;
+            m_pscan->Capture(&parsedClass);
+            m_pscan->Scan();
+
+            if (m_token.tk == tkID)
+            {
+                classHasName = true;
+            }
+
+            m_pscan->SeekTo(parsedClass);
+            pnode = ParseClassDecl<buildAST>(classHasName, nullptr, nullptr, nullptr);
+
+            if (buildAST)
+            {
+                AnalysisAssert(pnode != nullptr);
+                Assert(pnode->nop == knopClassDecl);
+
+                pnode->sxClass.SetIsDefaultModuleExport(true);
+            }
+
+            break;
+        }
+    case tkID:
+        // If we parsed an async token, it could either modify the next token (if it is a 
+        // function token) or it could be an identifier (let async = 0; export default async;).
+        // To handle both cases, when we parse an async token we need to keep the parser state
+        // and rewind if the next token is not function.
+        if (wellKnownPropertyPids.async == m_token.GetIdentifier(m_phtbl))
+        {
+            RestorePoint parsedAsync;
+            m_pscan->Capture(&parsedAsync);
+            m_pscan->Scan();
+            if (m_token.tk == tkFUNCTION)
+            {
+                // Token after async is function, consume the async token and continue to parse the 
+                // function as an async function.
+                flags |= fFncAsync;
+                goto LFunction;
+            }
+            // Token after async is not function, no idea what the async token is supposed to mean 
+            // so rewind and let the default case handle it.
+            m_pscan->SeekTo(parsedAsync);
+        }
+        goto LDefault;
+        break;
+    case tkFUNCTION:
+        {
+LFunction:
+            // We just parsed a function token but we need to figure out if the function 
+            // has an identifier name or not before we call the helper. 
+            RestorePoint parsedFunction;
+            m_pscan->Capture(&parsedFunction);
```

Comment:
It won’t throw if there’s no flag and no name. Anonymity is legal for non-declaration (function expression). I’d be curious to see what it is. If we can’t make that work for the initial check-in, can we open an issue to get rid of the rewind?

--Paul

From: Taylor Woll [mailto:notifications@github.com]
Sent: Monday, March 7, 2016 3:04 PM
To: Microsoft/ChakraCore <ChakraCore@noreply.github.com>
Cc: Paul Leathers <pleath@microsoft.com>
Subject: Re: [ChakraCore] Implement interpreter support for module import/exports (#457)


In lib/Parser/Parse.cpp<https://github.com/Microsoft/ChakraCore/pull/457#discussion_r55289195>:

> +                flags |= fFncAsync;

> +                goto LFunction;

> +            }

> +            // Token after async is not function, no idea what the async token is supposed to mean

> +            // so rewind and let the default case handle it.

> +            m_pscan->SeekTo(parsedAsync);

> +        }

> +        goto LDefault;

> +        break;

> +    case tkFUNCTION:

> +        {

> +LFunction:

> +            // We just parsed a function token but we need to figure out if the function

> +            // has an identifier name or not before we call the helper.

> +            RestorePoint parsedFunction;

> +            m_pscan->Capture(&parsedFunction);

With the declaration flag, it throws if the function doesn't have a name. With the no name flag, it throws if the function does have a name. I don't remember where it throws if you pass no flag at all, I think it throws if the function has no name (same as declaration flag).

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/ChakraCore/pull/457/files#r55289195>.